### ISSUE TITLE
PARTITION BY should support nested fields

### ISSUE THREAD

**Issue by rmoff** [confluentinc/ksql#2218]
```
ksql> DESCRIBE CUSTOMERS_STREAM_SRC;

Name                 : CUSTOMERS_STREAM_SRC
 Field   | Type
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 ROWTIME | BIGINT           (system)
 ROWKEY  | VARCHAR(STRING)  (system)
 BEFORE  | STRUCT<ID BIGINT, FIRST_NAME VARCHAR(STRING), LAST_NAME VARCHAR(STRING), EMAIL VARCHAR(STRING), GENDER VARCHAR(STRING), CLUB_STATUS VARCHAR(STRING), COMMENTS VARCHAR(STRING), CREATE_TS BIGINT, UPDATE_TS BIGINT>
 AFTER   | STRUCT<ID BIGINT, FIRST_NAME VARCHAR(STRING), LAST_NAME VARCHAR(STRING), EMAIL VARCHAR(STRING), GENDER VARCHAR(STRING), CLUB_STATUS VARCHAR(STRING), COMMENTS VARCHAR(STRING), CREATE_TS BIGINT, UPDATE_TS BIGINT>
 SOURCE  | STRUCT<VERSION VARCHAR(STRING), CONNECTOR VARCHAR(STRING), NAME VARCHAR(STRING), TS_MS BIGINT, TXID VARCHAR(STRING), SCN BIGINT, SNAPSHOT BOOLEAN>
 OP      | VARCHAR(STRING)
 TS_MS   | BIGINT
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
For runtime statistics and query details run: DESCRIBE EXTENDED <Stream,Table>;
```

Try to repartition stream based on AFTER->ID: 

```
ksql> CREATE STREAM CUSTOMERS_STREAM WITH (PARTITIONS=1) AS SELECT * FROM CUSTOMERS_STREAM_SRC PARTITION BY AFTER->ID;
line 1:224: mismatched input '->' expecting ';'
Caused by: org.antlr.v4.runtime.InputMismatchException
```

Workaround is a two-stream approach, to flatten the column and then re-partition: 

```
CREATE STREAM CUSTOMERS_STREAM1 WITH (PARTITIONS=1) AS SELECT AFTER->ID AS ID ,AFTER->FIRST_NAME AS FIRST_NAME ,AFTER->LAST_NAME AS LAST_NAME ,AFTER->EMAIL AS EMAIL ,AFTER->GENDER AS GENDER ,AFTER->CLUB_STATUS AS CLUB_STATUS ,AFTER->COMMENTS AS COMMENTS FROM CUSTOMERS_STREAM_SRC;

CREATE STREAM CUSTOMERS_STREAM WITH (PARTITIONS=1) AS SELECT * FROM CUSTOMERS_STREAM1 PARTITION BY ID;
```

**Comment by terryf82** [comment]
Looks like it is possible to get by with just one stream, as long as you select & alias the nested field first?

`CREATE STREAM CUSTOMERS_STREAM WITH (PARTITIONS=1) AS SELECT AFTER->ID AS ID, .... PARTITION BY ID;`

It even looks to work with CASE logic, too -

`CREATE STREAM CUSTOMERS_STREAM WITH (PARTITIONS=1) AS SELECT CASE WHEN OP = 'd' THEN BEFORE->ID ELSE AFTER->ID END AS ID, .... PARTITION BY ID;`

**Comment by agavra** [comment]
I believe this should be fixed by #4018 - I'll add a QTT test to verify this.

**Comment by agavra** [comment]
Confirmed this works:
```json
    {
      "name": "stream | initially set | partition by struct dereference expression | key in value | no aliasing",
      "statements": [
        "CREATE STREAM INPUT (foo VARCHAR, bar STRUCT<val VARCHAR>) WITH (kafka_topic='input_topic', key='foo', value_format='JSON');",
        "CREATE STREAM OUTPUT AS SELECT bar FROM INPUT PARTITION BY bar->val;"
      ],
      "inputs": [
        {"topic": "input_topic", "key": "1", "value": {"foo": "1", "bar": {"val":  "val"}}}
      ],
      "outputs": [
        {"topic": "OUTPUT", "key": "val", "value": {"BAR": {"VAL": "val"}}}
      ],
      "post": {
        "sources": [
          {"name": "OUTPUT", "type": "stream", "keyField": null}
        ]
      }
    },
```
I won't commit this because there is already a test for expression support and this doesn't add any coverage.
